File: cpp/91_Decode_Ways.py
class Solution:
    def dfs(self, i, s, memo):
        if i == len(s):
            return 1

        if s[i] == '0':
            return 0

        if memo[i] != -1:
            return memo[i]

        ways = self.dfs(i + 1, s, memo)

        if (
            i + 1 < len(s) and 
            (
                s[i] == '1' or 
                (s[i] == '2' and s[i + 1] <= '6')
            )
        ):
            ways += self.dfs(i + 2, s, memo)
        
        memo[i] = ways
        return ways


    def numDecodings(self, s: str) -> int:
        memo = [-1] * len(s)
        return self.dfs(0, s, memo)

File: cpp/91_Decode_Ways.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    int dfs(int i, string &s, vector<int> &memo) {
        if (i == s.size()) return 1;

        if (s[i] == '0') return 0;

        if (memo[i] != -1) return memo[i];

        int ways = dfs(i + 1, s, memo);

        if (i + 1 < s.size() &&
            (s[i] == '1' ||
            (s[i] == '2' && s[i + 1] <= '6')))
        {
            ways += dfs(i + 2, s, memo);
        }

        return memo[i] = ways;
    }

    int numDecodings(string s) {
        vector<int> memo(s.size(), -1);
        return dfs(0, s, memo);
    }
};
